src/main/java/com/aws/greengrass/componentmanager/plugins/DockerImageDownloader.java

```diff
@@ -188,4 +137,107 @@ public boolean canUnarchiveArtifact() {
+            if (credentialRefreshNeeded.get()) {
+                continue;
+            }
+
+            // Docker pull
+            run(() -> {
+                if (credentialsUsable(image)) {
+                    return dockerClient.pullImage(image);
+                } else {
+                    // Credentials have expired, re-fetch and login again
+                    logger.atInfo()
+                            .log("Registry credentials have expired, fetching fresh credentials and logging in again");
+                    credentialRefreshNeeded.set(true);
+                    return false;
+                }
+            }, "docker-pull-image", "Failed to download docker image");
+        } while (credentialRefreshNeeded.get());
+        // No file resources available since image artifacts are stored in docker's image store
+        return null;
+    }
+
+    /*
+     * Check if credentials are valid for private registries, always return true for public registries.
+     */
+    private boolean credentialsUsable(Image image) {
+        return !image.getRegistry().isPrivateRegistry() || image.getRegistry().getCredentials().stillValid();
+    }
+
+    private <T> void run(CrashableSupplier<T, Exception> task, String description, String message)
+            throws PackageDownloadException, InterruptedException {
+        try {
+            // Finite retry attempts for errors that are not due to connectivity issues and
+            // might need explicit intervention to recover from
+            RetryUtils.runWithRetry(finiteAttemptsRetryConfig, () -> RetryUtils
+                    // Indefinite retry for errors that are due to connectivity issues and can be
+                    // resolved when connectivity comes back
+                    .runWithRetry(infiniteAttemptsRetryConfig, () -> runWithConnectionErrorCheck(task), description,
+                            logger), description, logger);
+        } catch (InterruptedException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new PackageDownloadException(getErrorString(message), e);
+        }
+    }
+
+    private <T> T runWithConnectionErrorCheck(CrashableSupplier<T, Exception> task) throws Exception {
+        try {
+            return task.apply();
+        } catch (Exception e) {
+            // Since docker engine can throw service unavailable error that may or may not indicate a problem that
+            // can be fixed with retries, explicitly check if an error could be due to connectivity problem
+            // we infer that based on Mqtt connection, even though not perfect, it should accurately represent if
+            // device is having connectivity issues most of the times.
+            if (e instanceof DockerServiceUnavailableException && !mqttClient.getMqttOnline().get()) {
+                throw new ConnectionException(String.format("Device appears to be offline, should retry the task "), e);
```

Comment:
Is it possible to specify which task is being retried here?

Reply:
@saranyailla - The actual retry logic that wraps this does include the task description being retried, so we shouldn't need it in this message?

Reply:
@rbattle The formatter was a leftover, removed it now